lib/AST/DiagnosticConsumer.cpp

```diff
@@ -14,15 +14,148 @@
+
+llvm::SMLoc DiagnosticConsumer::getRawLoc(SourceLoc loc) {
+  return loc.Value;
+}
+
+FileSpecificDiagnosticConsumer::FileSpecificDiagnosticConsumer(
+    SmallVectorImpl<ConsumerPair> &consumers)
+  : SubConsumers(std::move(consumers)) {
+#ifndef NDEBUG
+  llvm::StringSet<> seenFiles;
+  for (const ConsumerPair &consumerPair : SubConsumers) {
+    if (consumerPair.first.empty())
+      continue;
+    bool isUnique = seenFiles.insert(consumerPair.first).second;
+    assert(isUnique &&
+           "duplicate file names given to FileSpecificDiagnosticConsumer");
+  }
+#endif
+}
+
+void
+FileSpecificDiagnosticConsumer::buildLocationToConsumerMap(SourceManager &SM) {
+  using MapEntry = std::pair<CharSourceRange, unsigned>;
+
+  // Look up each file's source range and add it to the "map" (to be sorted).
+  size_t i = 0;
+  for (const ConsumerPair &pair : SubConsumers) {
+    SWIFT_DEFER { ++i; };
+    if (pair.first.empty())
+      continue;
+
+    Optional<unsigned> bufferID = SM.getIDForBufferIdentifier(pair.first);
+    assert(bufferID.hasValue() && "consumer registered for unknown file");
+    CharSourceRange range = SM.getRangeForBuffer(bufferID.getValue());
+    LocationToConsumerMap.emplace_back(range, i);
+  }
+
+  // Sort the "map" by buffer /end/ location, for use with std::lower_bound
+  // later. (Sorting by start location would produce the same sort, since the
+  // ranges must not be overlapping, but since we need to check end locations
+  // later it's consistent to sort by that here.)
+  std::sort(LocationToConsumerMap.begin(), LocationToConsumerMap.end(),
+            [](const MapEntry &left, const MapEntry &right) -> bool {
+    auto compare = std::less<const char *>();
+    return compare(getRawLoc(left.first.getEnd()).getPointer(),
+                   getRawLoc(right.first.getEnd()).getPointer());
+  });
+
+  // Check that the ranges are non-overlapping.
+  assert(LocationToConsumerMap.end() ==
+           std::adjacent_find(LocationToConsumerMap.begin(),
+                              LocationToConsumerMap.end(),
+                              [](const MapEntry &left, const MapEntry &right) {
+                                return left.first.overlaps(right.first);
+                              }) &&
+         "overlapping ranges despite having distinct files");
+}
+
+DiagnosticConsumer *
+FileSpecificDiagnosticConsumer::consumerForLocation(SourceManager &SM,
+                                                    SourceLoc loc) const {
+  if (loc.isInvalid())
+    return nullptr;
+
+  // This map is generated on first use and cached, to allow the
+  // FileSpecificDiagnosticConsumer to be set up before the source files are
+  // actually loaded.
+  if (LocationToConsumerMap.empty()) {
+    auto *mutableThis = const_cast<FileSpecificDiagnosticConsumer*>(this);
+    mutableThis->buildLocationToConsumerMap(SM);
+  }
+
+  // "Find the first range whose end location is greater than or equal to
+  // 'loc'."
```

Comment:
Are the quotes around the comment a Doxygen thing?

Reply:
I was trying to imply that it was a sort of "translation" of the following invocation, rather than a higher-level operation. (Because as a higher-level operation, this would be a terrible comment. It doesn't tell you *why* it's looking for something `>=`.)

That didn't come across, so I'll think about rephrasing.

Reply:
That's the same name, no? The iterator for the range that might contain the location.